knownprojects_build/_python/dedup_units.py

```diff
@@ -0,0 +1,81 @@
+import pandas as pd
+
+from .utils import engine, psql_insert_copy
+
+
+def subtract_units(row, group):
+    higher_priority = group[group["source_id"] < row["source_id"]]
+    higher_priority_units = higher_priority["units_net"].sum()
+    row["units_net"] = row["units_net"] - higher_priority_units
+    if row["units_net"] < 0:
+        row["units_net"] = 0
+    return row
+
+
+def resolve_project(group):
+    if group.shape[0] > 1:
+        group = group.sort_values("source_id")
+        group = group.reset_index()
+        for index, row in group.iterrows():
+            group.iloc[index] = subtract_units(row, group)
+    return group
+
+
+def resolve_all_projects(df):
+    # Hierarchy for unit subtraction
+    hierarchy = {
+        "DOB": 1,
+        "HPD Projected Closings": 2,
+        "HPD RFPs": 3,
+        "EDC Projected Projects": 4,
+        "DCP Application": 5,
+        "Empire State Development Projected Projects": 6,
+        "Neighborhood Study Rezoning Commitments": 7,
+        "Neighborhood Study Projected Development Sites": 8,
+        "DCP Planner-Added Projects": 9,
+    }
+
+    df["source_id"] = df["source"].map(hierarchy)
+
+    # Subtract units within cluster based on hierarchy
+    print("Subtracting units within projcts based on source hierarchy...")
+    resolved = df.groupby(["project_id"], as_index=False).apply(resolve_project)
+    resolved = resolved[["source", "units_gross", "units_net", "project_id"]]
```

Comment:
We need to retain record_id in order to join the deduped units back onto the combined table.

Reply:
fixed